File: src/main/java/com/babylonml/backend/training/operations/LeakyLeRUFunction.java
package com.babylonml.backend.training.operations;

import com.babylonml.backend.cpu.TensorOperations;
import com.babylonml.backend.training.execution.TensorPointer;
import com.babylonml.backend.training.execution.TrainingExecutionContext;
import jdk.incubator.vector.FloatVector;
import jdk.incubator.vector.VectorOperators;
import jdk.incubator.vector.VectorSpecies;
import org.jspecify.annotations.NonNull;

import java.util.Objects;

public final class LeakyLeRUFunction extends AbstractOperation {
    private static final VectorSpecies<Float> SPECIES = FloatVector.SPECIES_PREFERRED;

    private final int @NonNull [] maxShape;

    private final boolean requiresDerivativeChainValue;
    private final float leakyLeRUASlope;

    private TensorPointer leftOperandResult;

    public LeakyLeRUFunction(float leakyLeRUASlope, Operation leftOperation) {
        this(null, leakyLeRUASlope, leftOperation);
    }

    public LeakyLeRUFunction(String name, float leakyLeRUASlope,
                             Operation leftOperation) {
        super(name, leftOperation, null);
        this.leakyLeRUASlope = leakyLeRUASlope;

        this.maxShape = TensorOperations.calculateMaxShape(leftOperation.getMaxResultShape(),
                leftOperation.getMaxResultShape());

        this.requiresDerivativeChainValue = leftOperation.requiresBackwardDerivativeChainValue();
    }

    @Override
    public int @NonNull [] getMaxResultShape() {
        return maxShape;
    }

    @Override
    public @NonNull TensorPointer forwardPassCalculation() {
        leftOperandResult = leftOperation.forwardPassCalculation();
        var result = executionContext.allocateForwardMemory(leftOperandResult.shape());

        var leftResultBuffer = leftOperandResult.buffer();
        var leftResultOffset = leftOperandResult.offset();

        var resultBuffer = result.buffer();
        var resultOffset = result.offset();

        var size = TensorOperations.stride(leftOperandResult.shape());
        var loopBound = SPECIES.loopBound(size);
        var zero = FloatVector.zero(SPECIES);

        for (int i = 0; i < loopBound; i += SPECIES.length()) {
            var va = FloatVector.fromArray(SPECIES, leftResultBuffer, leftResultOffset + i);
            var mask = va.compare(VectorOperators.LT, zero);
            var vc = va.mul(leakyLeRUASlope, mask);
            vc.intoArray(resultBuffer, resultOffset + i);
        }

        for (int i = loopBound; i < size; i++) {
            var leftValue = leftResultBuffer[leftResultOffset + i];
            resultBuffer[i + resultOffset] = leftValue > 0 ? leftValue : leakyLeRUASlope * leftValue;
        }

        return result;
    }

    @Override
    public @NonNull TensorPointer leftBackwardDerivativeChainValue() {
        Objects.requireNonNull(derivativeChainPointer);

        var leftOperandBuffer = leftOperandResult.buffer();
        var leftOperandOffset = leftOperandResult.offset();

        var derivativeChainBuffer = derivativeChainPointer.buffer();
        var derivativeChainOffset = derivativeChainPointer.offset();

        var result = executionContext.allocateBackwardMemory(leftOperandResult.shape());

        var resultBuffer = result.buffer();
        var resultOffset = result.offset();

        var size = TensorOperations.stride(leftOperandResult.shape());

        var loopBound = SPECIES.loopBound(size);
        var zero = FloatVector.zero(SPECIES);
        var slope = FloatVector.broadcast(SPECIES, leakyLeRUASlope);
        var one = FloatVector.broadcast(SPECIES, 1.0f);

        for (int i = 0; i < loopBound; i += SPECIES.length()) {
            var va = FloatVector.fromArray(SPECIES, leftOperandBuffer, leftOperandOffset + i);
            var mask = va.compare(VectorOperators.LT, zero);
            var vc = one.mul(slope, mask);

            var diff = FloatVector.fromArray(SPECIES, derivativeChainBuffer, derivativeChainOffset + i);
            vc = vc.mul(diff);

            vc.intoArray(resultBuffer, resultOffset + i);
        }

        for (int i = loopBound; i < size; i++) {
            resultBuffer[i + resultOffset] = (leftOperandBuffer[i + leftOperandOffset] > 0 ? 1.0f : leakyLeRUASlope) *
                    derivativeChainBuffer[i + derivativeChainOffset];
        }

        return result;
    }

    @Override
    public @NonNull TensorPointer rightBackwardDerivativeChainValue() {
        return TrainingExecutionContext.NULL;
    }

    @Override
    public boolean requiresBackwardDerivativeChainValue() {
        return requiresDerivativeChainValue;
    }


    @Override
    public int @NonNull [][] getForwardMemoryAllocations() {
        return new int[][]{
                maxShape
        };
    }

    @Override
    public int @NonNull [][] getBackwardMemoryAllocations() {
        return new int[][]{
                maxShape
        };
    }
}

File: src/main/kotlin/com/babylonml/backend/training/operations/LeakyLeRUFunction.kt
package com.babylonml.backend.training.operations

import com.babylonml.backend.cpu.TensorOperations
import com.babylonml.backend.training.execution.TensorPointer
import com.babylonml.backend.training.execution.TrainingExecutionContext
import it.unimi.dsi.fastutil.ints.IntImmutableList
import jdk.incubator.vector.FloatVector
import jdk.incubator.vector.VectorOperators
import jdk.incubator.vector.VectorSpecies

class LeakyLeRUFunction(
    name: String?, private val leakyLeRUASlope: Float,
    leftOperation: Operation
) : AbstractOperation(name, leftOperation, null) {
    override val maxResultShape: IntImmutableList =
        TensorOperations.calculateMaxShape(
            leftOperation.maxResultShape,
            leftOperation.maxResultShape
        )

    private var leftOperandResult: TensorPointer? = null

    constructor(leakyLeRUASlope: Float, leftOperation: Operation) : this(null, leakyLeRUASlope, leftOperation)

    override fun forwardPassCalculation(): TensorPointer {
        leftOperandResult = leftPreviousOperation!!.forwardPassCalculation()
        val result = executionContext.allocateForwardMemory(this, leftOperandResult!!.shape)

        val leftResultBuffer = leftOperandResult!!.buffer()
        val leftResultOffset = leftOperandResult!!.offset()

        val resultBuffer = result.buffer()
        val resultOffset = result.offset()

        val size = TensorOperations.stride(leftOperandResult!!.shape)
        val loopBound = SPECIES.loopBound(size)
        val zero = FloatVector.zero(SPECIES)

        run {
            var i = 0
            while (i < loopBound) {
                val va = FloatVector.fromArray(SPECIES, leftResultBuffer, leftResultOffset + i)
                val mask = va.compare(VectorOperators.LT, zero)
                val vc = va.mul(leakyLeRUASlope, mask)
                vc.intoArray(resultBuffer, resultOffset + i)
                i += SPECIES.length()
            }
        }

        for (i in loopBound until size) {
            val leftValue = leftResultBuffer[leftResultOffset + i]
            resultBuffer[i + resultOffset] = if (leftValue > 0) leftValue else leakyLeRUASlope * leftValue
        }

        return result
    }

    override fun leftBackwardDerivativeChainValue(): TensorPointer {
        val leftOperandBuffer = leftOperandResult!!.buffer()
        val leftOperandOffset = leftOperandResult!!.offset()

        val derivativeChainBuffer = derivativeChainPointer!!.buffer()
        val derivativeChainOffset = derivativeChainPointer!!.offset()

        val result = executionContext.allocateBackwardMemory(this, leftOperandResult!!.shape)

        val resultBuffer = result.buffer()
        val resultOffset = result.offset()

        val size = TensorOperations.stride(leftOperandResult!!.shape)

        val loopBound = SPECIES.loopBound(size)
        val zero = FloatVector.zero(SPECIES)
        val slope = FloatVector.broadcast(SPECIES, leakyLeRUASlope)
        val one = FloatVector.broadcast(SPECIES, 1.0f)

        run {
            var i = 0
            while (i < loopBound) {
                val va = FloatVector.fromArray(SPECIES, leftOperandBuffer, leftOperandOffset + i)
                val mask = va.compare(VectorOperators.LT, zero)
                var vc = one.mul(slope, mask)

                val diff = FloatVector.fromArray(SPECIES, derivativeChainBuffer, derivativeChainOffset + i)
                vc = vc.mul(diff)

                vc.intoArray(resultBuffer, resultOffset + i)
                i += SPECIES.length()
            }
        }

        for (i in loopBound until size) {
            resultBuffer[i + resultOffset] =
                (if (leftOperandBuffer[i + leftOperandOffset] > 0) 1.0f else leakyLeRUASlope) *
                        derivativeChainBuffer[i + derivativeChainOffset]
        }

        return result
    }

    override fun rightBackwardDerivativeChainValue(): TensorPointer {
        return TrainingExecutionContext.NULL
    }

    override val requiresBackwardDerivativeChainValue: Boolean by lazy {
        leftPreviousOperation!!.requiresBackwardDerivativeChainValue
    }

    override val forwardMemoryAllocations: List<IntImmutableList>
        get() = listOf(maxResultShape)

    override val backwardMemoryAllocations: List<IntImmutableList>
        get() = listOf(maxResultShape)

    companion object {
        private val SPECIES: VectorSpecies<Float> = FloatVector.SPECIES_PREFERRED
    }
}
